### PR TITLE
Binarydata fixes

### PR DESCRIPTION
**Pull request by erickvermot**
refers to https://github.com/kytos/python-openflow/issues/420

### CODE REVIEW

File: pyof/foundation/basic_types.py [hunk 1]
```diff
@@ -421,10 +421,23 @@ def __init__(self, value=b''):  # noqa
...
+    @staticmethod
+    def _pack_if_necessary(value):
+        if hasattr(value, 'pack') and callable(value.pack):
+            value = value.pack()
+        elif not value:
+            value = b''
+
+        if not isinstance(value, bytes):
+            msg = 'BinaryData value must contain bytes or have pack method; '
+            msg += 'Received type {} value: {}'.format(type(value), value)
```
Comment (cemsbr) [inline pyof/foundation/basic_types.py]: This is a better message for debugging. It is a good practice to put the value between double quotes so the user can visualize an empty value. When changing this, you can also fix a minor issue: use period or lowercase "Received".

File: pyof/foundation/basic_types.py [hunk 2]
```diff
@@ -421,10 +421,23 @@ def __init__(self, value=b''):  # noqa
...
-        if not isinstance(value, bytes):
-            raise ValueError('BinaryData must contain bytes.')
+        value = self._pack_if_necessary(value)
         super().__init__(value)
 
+    @staticmethod
+    def _pack_if_necessary(value):
+        if hasattr(value, 'pack') and callable(value.pack):
+            value = value.pack()
+        elif not value:
```
Comment (cemsbr) [inline pyof/foundation/basic_types.py]: `not value` is dangerous. Here, `pack(42)` raises, but `pack(0)` doesn't. It would be more consistent to explicitly write `value is None`.

File: pyof/foundation/basic_types.py [hunk 3]
```diff
@@ -412,7 +412,7 @@ class BinaryData(GenericType):
     return the size of the instance using Python's :func:`len`.
     """
 
-    def __init__(self, value=b''):  # noqa
+    def __init__(self, value=None):  # noqa
```
Comment (cemsbr) [inline pyof/foundation/basic_types.py]: Accepting None will be useful to change body's default to None. Please, add a test for this behavior.

File: pyof/foundation/basic_types.py (hunk 2 above)
Comment (cemsbr) [inline pyof/foundation/basic_types.py]: Please, write tests for such cases.

File: pyof/foundation/basic_types.py [hunk 4]
```diff
@@ -421,10 +421,23 @@ def __init__(self, value=b''):  # noqa
         Raises:
             ValueError: If given value is not bytes.
         """
-        if not isinstance(value, bytes):
-            raise ValueError('BinaryData must contain bytes.')
+        value = self._pack_if_necessary(value)
         super().__init__(value)
 
+    @staticmethod
+    def _pack_if_necessary(value):
```
Comment (cemsbr) [inline pyof/foundation/basic_types.py]: I believe the `_if_necessary` suffix is an implementation detail that shouldn't be part of the method's name in this case. `_pack(value)` together with `@staticmethod` is enough for understanding its functionality.

File: pyof/foundation/basic_types.py (hunk 2 above)
Comment (erickvermot) [inline pyof/foundation/basic_types.py]: I wrote this not for the case where value is None, but for the cases where it is an empty typle or list for example, which is a behavior it already had. I do not think it is a good idea, but kept it simply because it was there, and people were using it.
I'll remove it.

File: pyof/foundation/basic_types.py (hunk 4 above)
Comment (erickvermot) [inline pyof/foundation/basic_types.py]: I think leaving only _pack makes one think that it will pack anyway; furthermore, the ``_pack`` method was used for another purpose in the original PR from which this commit came.

File: pyof/foundation/basic_types.py (hunk 3 above)
Comment (erickvermot) [inline pyof/foundation/basic_types.py]: sorry, but I don't see what kind of test do you expect...

File: pyof/foundation/basic_types.py (hunk 2 above)
Comment (erickvermot) [inline pyof/foundation/basic_types.py]: now we'll have to wait https://github.com/kytos/python-openflow/pull/417

File: pyof/foundation/basic_types.py (hunk 3 above)
Comment (cemsbr) [inline pyof/foundation/basic_types.py]: By "test", I mean unit tests in tests folder. In this case, asserting that the class behaves correctly with a None value.

Comment (erickvermot) [inline pyof/foundation/basic_types.py]: this is already covered here https://github.com/kytos/python-openflow/blob/master/tests/test_foundation/test_basic_types.py#L169 , isn't it?
I don't think adding ``None`` instead of leaving the default initialization is of much use... All other ``GenericTypes`` accept ``None``, and we are testing none of them.
If you think it is important to add tests for ``GenericType`` accepting ``None`` explicitly, maybe we should write one that does the test for all of them.

**Comment by erickvermot** [comment]
https://github.com/kytos/python-openflow/pull/421#discussion_r129399116

File: pyof/foundation/basic_types.py (hunk 3 above)
Comment (cemsbr) [inline pyof/foundation/basic_types.py]: That test is different and does not cover this case. We need a new test to make it explicit you need the None behavior and that it works as expected.

Writing tests to other structures is out of the scope of this issue/PR.

File: pyof/foundation/basic_types.py [hunk 5]
```diff
@@ -421,10 +421,23 @@ def __init__(self, value=b''):  # noqa
         Raises:
             ValueError: If given value is not bytes.
         """
-        if not isinstance(value, bytes):
-            raise ValueError('BinaryData must contain bytes.')
+        value = self._pack_value(value)
         super().__init__(value)
 
+    @staticmethod
+    def _pack_value(value):
```
Comment (cemsbr) [inline pyof/foundation/basic_types.py]: `_pack_value` is redundant. Just `_pack` is a better name.

File: pyof/foundation/basic_types.py [hunk 6]
```diff
@@ -421,10 +421,23 @@ def __init__(self, value=b''):  # noqa
...
+        if hasattr(value, 'pack') and callable(value.pack):
+            value = value.pack()
+        elif value is None:
+            value = b''
+
+        if not isinstance(value, bytes):
+            msg = 'BinaryData value must contain bytes or have pack method; '
+            msg += 'Received type {} value: "{}"'.format(type(value), value)
```
Comment (cemsbr) [inline pyof/foundation/basic_types.py]: The review about this sentence wasn't solved.

File: pyof/foundation/basic_types.py (hunk 5 above)
Comment (erickvermot) [inline pyof/foundation/basic_types.py]: I am not sure you read my reply... It looks like you didn't.
I changed it anyway...

**Comment by erickvermot** [comment]
@cemsbr As I mentioned in the previous comment (https://github.com/kytos/python-openflow/pull/421#issuecomment-318060563), because of the changes you requested, it now depends on #417

**Comment by erickvermot** [comment]
since #417 was accepted, this should pass the tests now.

**Comment by beraldoleal** [comment]
@macartur can you please try to understand this issue and make sure that all comments / changes requests are ok ? Thanks a lot.

**Comment by diraol** [comment]
Changes contemplated on #445
